highfidelity/hifi
# Re-parent entities after duplicating and duplicate children with parents

When duplicating entities, store the entities that had parents and will need re-parenting, and also store the mapping of original entity ID to new duplicated entity ID. After duplicating, for each newly duplicated entity whose parent was also duplicated, re-parent them to the new duplicate of the original parent entity. Also ensure any unselected children of selected parent entities get duplicated.

Fixes: https://highfidelity.manuscript.com/f/cases/13845

Test Plan:

The new duplicating parent/child rules will mainly follow these guidelines:
-Duplicating an entity who is a parent will always duplicate with all it's children (including children of children and so on downward), whether those children are selected at time of duplication or not
-Duplicating an entity who is a child will only duplicate with it's parent if it's parent is also selected at time of duplication
-Duplicating an entity who is a child will cause the new duplicate to be parented to the parent of the original entity being duplicated

Test Steps:
-Enter Interface in desktop mode
-Open Create and create several entities of any type
-Select any entity and verify when holding Alt to drag it that a single-selected entity copies as expected
-Select any entity to find and copy it's ID in the properties tab
-Select any other entity and assign it's parent ID to the copied ID above
-Select both entities and hold Alt to drag and duplicate them, verifying the same parent/child relationship is maintained between the new duplicated entities
-Select only the child entity and hold Alt to drag and duplicate it, verifying the original entity and the new duplicated entity are parented to the same parent
-Parent more entities together linearly (i.e. entity D parented to entity C parented to entity B parented to entity A)
-Select the top-level parent (entity A in the example in the step above) and hold Alt to drag and duplicate it, verifying any unselected children of the original selected entity are also duplicated and the same parent/child relationship is maintained between the new set of duplicated entities and the original set of entities
-Select an entity in the middle of the linear hierarchy from above, and hold Alt to drag and duplicate it, verifying any unselected children (and not the parent) of the original selected entity are also duplicated, and the same parent/child relationship is maintained between the new set of duplicated entities and the original set of entities which includes parenting the duplicate of the original selected entity to the same parent as the original
-Select all of the linearly parented entities above and hold Alt to drag and duplicate them, verifying the same parent/child relationship is maintained between the new set of duplicated entities and the original set of entities
-Parent more entities together in a random/non-linear fashion (i.e. entity E parented to entity A, entity D parented to entity A, entity C parented to entity B, entity B parented to entity A, etc.)
-Select an entity or entities anywhere in the hierarchy and hold Alt to drag and duplicate it/them, verifying any unselected children (and not the parent) of the original selected entity are also duplicated, and the same parent/child relationship is maintained between the new set of duplicated entities and the original set of entities which includes parenting the duplicate of the original selected entity to the same parent as the original
-Select all the non-linear parented entities above and hold Alt to drag and duplicate them, verifying the same parent/child relationship is maintained between the new set of duplicated entities and the original set of entities
